### PR TITLE
CDI interceptors on MP Rest Client interfaces

### PR DESCRIPTION
**Pull request by andymc12**
Enables CDI interceptors to be invoked with their annotations are placed on the MP Rest Client interface (type or method level).

### CODE REVIEW

**Comment by andymc12** [comment]
#build

**Comment by andymc12** [comment]
#build

**Comment by andymc12** [comment]
Failures are unrelated (and already have defects/issues associated with them).

**Comment by andymc12** [comment]
@jim-krueger good idea!  in this case, all of the new files are actually new files - not pulled in from CXF - though I will most likely contribute them to CXF in the future.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/CDIInterceptorWrapperImpl.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                Annotation[] interceptorBindings = merge(methodBindings, classBindings);
+
+                List<Interceptor<?>> interceptors = new ArrayList<>(beanManager.resolveInterceptors(InterceptionType.AROUND_INVOKE, interceptorBindings));
+                if (LOG.isLoggable(Level.FINEST)) {
+                    LOG.finest("Resolved interceptors from beanManager, " + beanManager + ":" + interceptors);
+                }
+                // hack for FT interceptors
+                if (FT_ANNO_CLASS != null && containsFTannotation(interceptorBindings)) {
+                    interceptors.addAll(0, getFTInterceptor(beanManager));
+                }
```
Comment (Azquelt) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/CDIInterceptorWrapperImpl.java]: Can we add the `FT_ANNO_CLASS` to the `methodBindings` instead? Otherwise the ordering of interceptors will be wrong.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/bnd.overrides [hunk 2]
```diff
@@ -50,9 +50,14 @@ Import-Package: \
  com.ibm.websphere.ras,\
  com.ibm.wsspi.classloading,\
  com.ibm.ws.cdi.extension;resolution:=optional,\
+ com.ibm.ws.ffdc,\
  com.ibm.ws.jaxrs20.client.configuration
 # org.apache.cxf.jaxrs.provider.jsrjsonp;version="[3.1,4)",\
 
+DynamicImport-Package: \
+ javax.interceptor,\
+ com.ibm.ws.microprofile.faulttolerance.cdi
```
Comment (Azquelt) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/bnd.overrides]: Why are these needed?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    boolean hasNextInterceptor() {
+        return index < interceptorInvokers.size();
+    }
+
+    protected Object invokeNextInterceptor() throws Exception {
+        int oldIndex = index;
+        try {
+            return interceptorInvokers.get(index++).invoke(this);
```
Comment (Azquelt) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java]: This will go hilariously wrong if there are interceptors after the fault tolerance interceptor.
* `Retry` can cause later interceptors to be called more than once
* `Asynchronous` can cause later interceptors to be run on a different thread, possibly after all the earlier interceptors have completed.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Override
+    public Object proceed() throws Exception {
+        try {
+            if (hasNextInterceptor()) {
+                return invokeNextInterceptor();
+            } else {
+                return interceptorChainCompleted();
+            }
+        } catch (InvocationTargetException e) {
```
Comment (Azquelt) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java]: Why do we need to treat `InvocationTargetException` specially here?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/bnd.overrides (hunk 2 above)
Comment (andymc12) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/bnd.overrides]: `javax.interceptor` is required for the `InvocationContext` - used in InterceptorInvoker and MPRestClientInvocationContextImpl.

`com.ibm.ws.microprofile.faulttolerance.cdi` is used to load the `FaultTolerance` annotation class.

These classes should only be required when using MP Rest Client _and_ MP Fault Tolerance, so that's why I used DynamicImport rather than normal Import.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java (hunk 4 above)
Comment (andymc12) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java]: We don't - I think that was leftover code from some reflection stuff I was doing but later refactored.  Will remove.  Good catch!

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java (hunk 3 above)
Comment (andymc12) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/MPRestClientInvocationContextImpl.java]: I can add a comment to that effect here - but I think the FT interceptor should always be the last one invoked due to the ordering imposed in CDIInterceptorWrapperImpl.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/CDIInterceptorWrapperImpl.java (hunk 1 above)
Comment (andymc12) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.mp.client.3.2/src/org/apache/cxf/microprofile/client/cdi/CDIInterceptorWrapperImpl.java]: Done - good idea, thanks!

**Comment by andymc12** [comment]
Rebased - new #build

**Comment by andymc12** [comment]
Closing, as this is resolved in CXF for MP Rest Client 1.2.
